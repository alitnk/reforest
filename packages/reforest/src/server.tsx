import * as React from "react"
import { suspend } from "suspend-react"

import { ComputedDataContext, TreeMapContext } from "./contexts"
import { isServer, sortMapByIndexPath } from "./utils"

const DATA_ID = "__REFOREST_DATA__"

/** Stringify a tree map. */
export function stringifyTreeMap(treeMap: Map<string, any>) {
  let allData = {}

  treeMap.forEach((data, key) => {
    allData[key] = data
  })

  return JSON.stringify(allData)
}

/** Wraps a tree with a Map for gathering indexed data on the server. */
export function createTreeProvider(initialEntries?: [string, any][]) {
  const treeComputedData = new Map<string, any>(initialEntries)

  function TreeProvider(props: { children: React.ReactNode }) {
    return (
      <ComputedDataContext.Provider value={treeComputedData}>
        {props.children}
      </ComputedDataContext.Provider>
    )
  }

  return {
    TreeProvider,
    treeComputedData,
    stringifyTreeComputedData: () => stringifyTreeMap(treeComputedData),
    getInitializerScript: () =>
      `<script id="${DATA_ID}" type="application/json">${stringifyTreeMap(
        treeComputedData
      )}</script>`,
  }
}

/** Track initial component renders. */
let globalResolves: any[] = []
let globalTimeoutId: ReturnType<typeof setTimeout>

/** Compute data based on all collected tree data. */
export function useServerComputedData<TreeValue extends any, ComputedTreeValue extends any>(
  treeId: string,
  computeData?: (treeMap: Map<string, TreeValue>, treeId: string) => ComputedTreeValue
) {
  const treeMap = React.useContext(TreeMapContext)
  const treeComputedData = React.useContext(ComputedDataContext)

  /** Use Suspense to re-render the component before committing the final props on the server. */
  const isServerWithComputedData = isServer && treeMap && computeData !== undefined
  let serverComputedData: any = null

  if (isServerWithComputedData) {
    serverComputedData = suspend(() => {
      return new Promise((resolve) => {
        /** Keep clearing timeout until the last component renders. */
        clearTimeout(globalTimeoutId)

        /** Store all of the promises to compute. */
        globalResolves.push(() => {
          const sortedTreeMap = sortMapByIndexPath(treeMap)
          const computedData = computeData(sortedTreeMap, treeId)

          resolve(computedData)
        })

        /** Push to the end of the event stack to allow all leaf components to initially render. */
        globalTimeoutId = setTimeout(() => {
          /** Resolve all of the leaf promises now that we have stored and computed all data. */
          globalResolves.forEach((resolve) => resolve())
          globalResolves = []
        })
      })
    }, [treeId])

    /** Store computed data so it can be injected on the server. */
    treeComputedData?.set(treeId, serverComputedData)
  } else {
    serverComputedData = treeComputedData?.get(treeId)
  }

  return serverComputedData
}
