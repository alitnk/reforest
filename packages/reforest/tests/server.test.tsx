/** @jest-environment node */
import * as React from "react"
import * as ReactDOMServer from "react-dom/server"
import { Writable } from "stream"

import { createTreeProvider, useTree, useTreeData } from "../index"
import { App } from "./App"

/** Simple render function to mock what renderToPipeableStream does. */
function render(element: React.ReactNode) {
  return new Promise((resolve) => {
    let isReady = false
    const response = new (class extends Writable {
      _write(chunk, _, next) {
        if (isReady) {
          resolve(chunk.toString())
        }
        next()
      }
    })()
    const stream = ReactDOMServer.renderToPipeableStream(element, {
      onShellReady() {
        stream.pipe(response)
      },
      onAllReady() {
        isReady = true
      },
    })
  })
}

test("computed data renders on server", async () => {
  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const data = useTreeData(
      React.useMemo(() => ({ value }), [value]),
      (tree) => {
        if (tree.map) {
          return tree.map.size
        }
        return 0
      }
    )
    const tree = useTree(children)

    return (
      <div data-testid={data?.indexPathString}>
        {data!.computed} {tree.children}
      </div>
    )
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    return tree.children
  }

  const { TreeProvider, stringifyTreeCollection } = createTreeProvider()
  const renderedString = await render(
    <React.Suspense fallback={null}>
      <TreeProvider>
        <ItemList>
          <Item value="apple">Apple</Item>
          <Item value="orange">Orange</Item>
          <Item value="banana">Banana</Item>
        </ItemList>
      </TreeProvider>
    </React.Suspense>
  )

  expect(renderedString).toMatchSnapshot()
  expect(stringifyTreeCollection()).toMatchSnapshot()
})

test("changing rendered elements based on computed data", async () => {
  function Box({ id }: { id: string }) {
    const data = useTreeData(
      React.useMemo(() => ({ id }), [id]),
      (tree, generatedId) => {
        const ids = new Set()
        let shouldRender = false

        tree.map?.forEach(({ id }, generatedIdToCompare) => {
          const isSameId = generatedId === generatedIdToCompare
          const hasId = ids.has(id)

          if (isSameId) {
            shouldRender = !hasId
          }

          if (!hasId) {
            ids.add(id)
          }
        })

        return shouldRender
      }
    )

    return data?.computed ? <div id={id} /> : null
  }

  function Parent({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    return tree.children
  }

  const { TreeProvider } = createTreeProvider()
  const renderedString = await render(
    <React.Suspense fallback={null}>
      <TreeProvider>
        <Parent>
          <Box id="a" />
          <Box id="b" />
          <Box id="c" />
          <Box id="b" />
        </Parent>
      </TreeProvider>
    </React.Suspense>
  )

  expect(renderedString).toMatchSnapshot()
})

test("tree collection", async () => {
  const { TreeProvider, stringifyTreeCollection } = createTreeProvider()

  const renderedString = await render(
    <TreeProvider>
      <App />
    </TreeProvider>
  )

  expect(renderedString).toMatchSnapshot()
  expect(stringifyTreeCollection()).toMatchSnapshot()
})
