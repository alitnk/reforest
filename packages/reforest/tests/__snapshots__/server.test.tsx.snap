// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changing rendered elements based on computed data 1`] = `"<div hidden id=\\"S:1\\"></div><script>function $RS(a,b){a=document.getElementById(a);b=document.getElementById(b);for(a.parentNode.removeChild(a);a.firstChild;)b.parentNode.insertBefore(a.firstChild,b);b.parentNode.removeChild(b)};$RS(\\"S:1\\",\\"P:1\\")</script><script>function $RC(a,b){a=document.getElementById(a);b=document.getElementById(b);b.parentNode.removeChild(b);if(a){a=a.previousSibling;var f=a.parentNode,c=a.nextSibling,e=0;do{if(c&&8===c.nodeType){var d=c.data;if(\\"/$\\"===d)if(0===e)break;else e--;else\\"$\\"!==d&&\\"$?\\"!==d&&\\"$!\\"!==d||e++}d=c.nextSibling;f.removeChild(c);c=d}while(c);for(;b.firstChild;)f.insertBefore(b.firstChild,c);a.data=\\"$\\";a._reactRetry&&a._reactRetry()}};$RC(\\"B:0\\",\\"S:0\\")</script>"`;

exports[`computed data renders on server 1`] = `"<div hidden id=\\"S:1\\"><div data-testid=\\"0\\">3<!-- --> <!-- -->Apple</div></div><script>function $RS(a,b){a=document.getElementById(a);b=document.getElementById(b);for(a.parentNode.removeChild(a);a.firstChild;)b.parentNode.insertBefore(a.firstChild,b);b.parentNode.removeChild(b)};$RS(\\"S:1\\",\\"P:1\\")</script><div hidden id=\\"S:2\\"><div data-testid=\\"1\\">3<!-- --> <!-- -->Orange</div></div><script>$RS(\\"S:2\\",\\"P:2\\")</script><div hidden id=\\"S:3\\"><div data-testid=\\"2\\">3<!-- --> <!-- -->Banana</div></div><script>$RS(\\"S:3\\",\\"P:3\\")</script><script>function $RC(a,b){a=document.getElementById(a);b=document.getElementById(b);b.parentNode.removeChild(b);if(a){a=a.previousSibling;var f=a.parentNode,c=a.nextSibling,e=0;do{if(c&&8===c.nodeType){var d=c.data;if(\\"/$\\"===d)if(0===e)break;else e--;else\\"$\\"!==d&&\\"$?\\"!==d&&\\"$!\\"!==d||e++}d=c.nextSibling;f.removeChild(c);c=d}while(c);for(;b.firstChild;)f.insertBefore(b.firstChild,c);a.data=\\"$\\";a._reactRetry&&a._reactRetry()}};$RC(\\"B:0\\",\\"S:0\\")</script>"`;

exports[`computed data renders on server 2`] = `"{\\"R0\\":{\\"R3\\":{\\"generatedId\\":\\"R3\\",\\"indexPathString\\":\\"0\\",\\"value\\":\\"apple\\"},\\"R5\\":{\\"generatedId\\":\\"R5\\",\\"indexPathString\\":\\"1\\",\\"value\\":\\"orange\\"},\\"R7\\":{\\"generatedId\\":\\"R7\\",\\"indexPathString\\":\\"2\\",\\"value\\":\\"banana\\"}}}"`;

exports[`tree collection 1`] = `"<div hidden id=\\"S:1\\"><div style=\\"width:200px;height:133.33333333333334px;top:201px;left:134.33333333333334px\\">Box</div></div><script>function $RS(a,b){a=document.getElementById(a);b=document.getElementById(b);for(a.parentNode.removeChild(a);a.firstChild;)b.parentNode.insertBefore(a.firstChild,b);b.parentNode.removeChild(b)};$RS(\\"S:1\\",\\"P:1\\")</script><script>function $RC(a,b){a=document.getElementById(a);b=document.getElementById(b);b.parentNode.removeChild(b);if(a){a=a.previousSibling;var f=a.parentNode,c=a.nextSibling,e=0;do{if(c&&8===c.nodeType){var d=c.data;if(\\"/$\\"===d)if(0===e)break;else e--;else\\"$\\"!==d&&\\"$?\\"!==d&&\\"$!\\"!==d||e++}d=c.nextSibling;f.removeChild(c);c=d}while(c);for(;b.firstChild;)f.insertBefore(b.firstChild,c);a.data=\\"$\\";a._reactRetry&&a._reactRetry()}};$RC(\\"B:0\\",\\"S:0\\")</script>"`;

exports[`tree collection 2`] = `"{\\"R0\\":{\\"R3H1\\":{\\"generatedId\\":\\"R3H1\\",\\"indexPathString\\":\\"0\\",\\"type\\":\\"row\\",\\"id\\":null,\\"width\\":\\"auto\\",\\"height\\":\\"auto\\",\\"weight\\":1,\\"children\\":[]},\\"Rf\\":{\\"generatedId\\":\\"Rf\\",\\"indexPathString\\":\\"0.0\\",\\"type\\":\\"space\\",\\"width\\":\\"fill\\",\\"height\\":\\"fill\\",\\"id\\":null,\\"weight\\":1},\\"RnH1\\":{\\"generatedId\\":\\"RnH1\\",\\"indexPathString\\":\\"0.1\\",\\"type\\":\\"column\\",\\"id\\":null,\\"width\\":\\"auto\\",\\"height\\":\\"auto\\",\\"weight\\":1},\\"R3n\\":{\\"generatedId\\":\\"R3n\\",\\"indexPathString\\":\\"0.1.0\\",\\"type\\":\\"space\\",\\"width\\":\\"fill\\",\\"height\\":\\"fill\\",\\"id\\":null,\\"weight\\":1},\\"R5n\\":{\\"generatedId\\":\\"R5n\\",\\"indexPathString\\":\\"0.1.1\\",\\"type\\":\\"node\\",\\"width\\":\\"fill\\",\\"height\\":\\"fill\\",\\"id\\":null,\\"weight\\":1},\\"R7n\\":{\\"generatedId\\":\\"R7n\\",\\"indexPathString\\":\\"0.1.2\\",\\"type\\":\\"space\\",\\"width\\":\\"fill\\",\\"height\\":\\"fill\\",\\"id\\":null,\\"weight\\":1},\\"Rv\\":{\\"generatedId\\":\\"Rv\\",\\"indexPathString\\":\\"0.2\\",\\"type\\":\\"space\\",\\"width\\":\\"fill\\",\\"height\\":\\"fill\\",\\"id\\":null,\\"weight\\":1}}}"`;
